MobiFlight/MidiBoard/MidiBoardManager.cs

```diff
@@ -0,0 +1,273 @@
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Windows.Forms;
+
+namespace MobiFlight
+{
+    public class MidiBoardManager
+    {
+        public event EventHandler Connected;
+        public event ButtonEventHandler OnButtonPressed;      
+        private readonly List<MidiBoard> MidiBoards = new List<MidiBoard>();
+        private readonly List<MidiBoard> BoardsToBeRemoved = new List<MidiBoard>();
+        private readonly Dictionary<string, MidiBoardDefinition> Definitions = new Dictionary<string, MidiBoardDefinition>();
+        private readonly Timer ProcessTimer = new System.Windows.Forms.Timer();
+        private int CheckAttachedRemovedCounter = 0;
+
+        public MidiBoardManager ()
+        {
+            LoadDefinitions();
+            ProcessTimer.Interval = 50;
+            ProcessTimer.Tick += ProcessTimer_Tick;
+        }
+
+        private List<string> CollectErrorsInInputDefinition(MidiBoardDefinition def)
+        {
+            List<string> errorMessages = new List<string>();
+            if (def.Inputs == null)
+                errorMessages.Add($"No input definitions.");
+            else
+            {
+                foreach (var item in def.Inputs)
+                {
+                    if (item.LabelIds.Length != item.MessageIds.Length)
+                        errorMessages.Add($"{item.Label}: Number of LabelIds unequal to number of MessageIds");
+                    if (!item.Label.Contains("%"))
+                        errorMessages.Add($"{item.Label}: Label does not contain % placeholder.");
+                }
+            }
+            return errorMessages;
+        }
+
+        private List<string> CollectErrorsInOutputDefinition(MidiBoardDefinition def)
+        {
+            List<string> errorMessages = new List<string>();
+            if (def.Outputs != null)
+            {
+                foreach (var item in def.Outputs)
+                {
+                    if (item.LabelIds.Length != item.MessageIds.Length)
+                        errorMessages.Add($"{item.Label}: Number of LabelIds unequal to number of MessageIds");
+                    if (!item.Label.Contains("%"))
+                        errorMessages.Add($"{item.Label}: Label does not contain % placeholder.");
+                    if (!string.IsNullOrEmpty(item.RelatedInputLabel))
+                    {
+                        if (item.RelatedIds.Length != item.MessageIds.Length)
+                            errorMessages.Add($"{item.Label}: Number of RelatedIds unequal to number of MessageIds");
+                        if (!item.RelatedInputLabel.Contains("%"))
+                            errorMessages.Add($"{item.Label}: RelatedInputLabel does not contain % placeholder.");
+                    }
+                }
+            }
+            return errorMessages;
+        }
+
+
+        private bool IsErrorInDefinition(MidiBoardDefinition def)
```

Comment:
It's a little weird to have a function called `IsErrorInDefinition()` both test if there is an error (the return statement) *and* log all the error messages.

Reply:
Agree that should be handled and named differently.